src/Symfony/Component/Form/Extension/DependencyInjection/DependencyInjectionExtension.php

```diff
@@ -82,8 +114,13 @@ public function getTypeGuesser()
             $this->guesserLoaded = true;
             $guessers = array();
 
-            foreach ($this->guesserServiceIds as $serviceId) {
-                $guessers[] = $this->container->get($serviceId);
+            foreach ($this->guesserServices as $serviceId => $service) {
+                if (is_string($service)) {
+                    $serviceId = $service;
+                    $service = $this->typeContainer->get($serviceId);
+                }
+
+                $guessers[] = $service;
```

Comment:
> IMHO the extension is in charge of building the whole thing

Too many responsibilities to me.

> and the use of the FormTypeGuesserChain is an implementation detail. There is no need to pass the guesser chain as a dependency.

The fact the guessers are multiple services is a detail IMO. At the end we have only one `FormTypeGuesserInterface $guesser`, the class could be usable with an iterable or a simple guesser, which looks better to me.

> There is an issue with FormTypeGuesserChain, though... its constructor. It iterates over the guessers and this would trigger the loading too soon anyway.

Just validation/merge, can be moved at compile time I think.